File: customlayoutmanagerdemo/src/main/java/com/example/customlayoutmanagerdemo/MainActivity.java
package com.example.customlayoutmanagerdemo;

import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.PagerSnapHelper;
import androidx.recyclerview.widget.RecyclerView;
import androidx.recyclerview.widget.SnapHelper;

import android.os.Bundle;

import com.example.customlayoutmanagerdemo.adapter.MyAdapter;
import com.example.customlayoutmanagerdemo.itemdecoration.HeadAndTailSpaceDecoration;
import com.example.customlayoutmanagerdemo.layoutmanager.MyLayoutManager;

public class MainActivity extends AppCompatActivity {
    private RecyclerView recycler;
    private SnapHelper mSnapHelper;
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        recycler = findViewById(R.id.recycler);
        mSnapHelper = new PagerSnapHelper();
        mSnapHelper.attachToRecyclerView(recycler);
        recycler.addItemDecoration(new HeadAndTailSpaceDecoration(ItemConfigure.getDefault().getBorderItemHorizontalMargin()));
        recycler.setLayoutManager(new MyLayoutManager(this, LinearLayoutManager.HORIZONTAL, false));
        MyAdapter myAdapter = new MyAdapter(this);
        recycler.setAdapter(myAdapter);
    }
}

File: concatadapterdemo/src/main/java/com/example/concatadapterdemo/MainActivity.kt
package com.example.concatadapterdemo

import android.content.Intent
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import kotlinx.android.synthetic.main.activity_main.*
import java.util.*

class MainActivity : AppCompatActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        btn_open_concat_page.setOnClickListener {
            openActivity(ConcatAdapterActivity::class.java)
        }
        btn_open_vlayout_page.setOnClickListener {
            openActivity(VLayoutActivity::class.java)
        }
    }

    fun openActivity(className: Class<*>){
        val intent = Intent(this, className)
        startActivity(intent)
    }
}